test/fetchzip.sh

```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+DIR=$(dirname "${BASH_SOURCE[0]}")
+
+. "$DIR"/common.sh
+
+PORT=$(($$ + 1024)) # PID + 1024
+
+dub remove gitcompatibledubpackage --non-interactive --version=* 2>/dev/null || true
+
+"$DUB" build --single "$DIR"/test_registry.d
+"$DIR"/test_registry --folder="$DIR/issue1336-registry" --port=$PORT &
+PID=$!
+sleep 0.2
+trap 'kill $PID 2>/dev/null || true' exit
+
+echo "Trying to download gitcompatibledubpackage (1.0.4)"
+timeout 1s "$DUB" fetch gitcompatibledubpackage --version=1.0.4 --skip-registry=all --registry=http://localhost:$PORT
+if [ $? -eq 124 ]; then
+    die 'Fetching from responsive registry should not time-out.'
+fi
+dub remove gitcompatibledubpackage --non-interactive --version=1.0.4
+
+echo "Downloads should be retried when the zip is corrupted - gitcompatibledubpackage (1.0.3)"
+zipOut=$(! timeout 1s "$DUB" fetch gitcompatibledubpackage --version=1.0.3 --skip-registry=all --registry=http://localhost:$PORT 2>&1)
+rc=$?
+
+if ! zipCount=$(grep -Fc 'Failed to extract zip archive' <<<"$zipOut") || [ "$zipCount" -lt 3 ] ; then
+    echo '========== +Output was ==========' >&2
+    echo "$zipOut" >&2
+    echo '========== -Output was ==========' >&2
+    die 'DUB should have tried to download the zip archive multiple times.'
+elif [ $rc -eq 124 ]; then
+    die 'DUB timed out unexpectedly.'
+fi
+if dub remove gitcompatibledubpackage --non-interactive --version=* 2>/dev/null; then
+    die 'DUB should not have installed a broken package.'
+fi
+
+echo "HTTP status errors on downloads should be retried - gitcompatibledubpackage (1.0.2)"
+retryOut=$(! timeout 1s "$DUB" fetch gitcompatibledubpackage --version=1.0.2 --skip-registry=all --registry=http://localhost:$PORT --vverbose 2>&1)
```

Comment:
The Travis log shows an error and `retryCount` of `1` -  despite the log clearly showing "Bad Gateway" three times:

```
Downloading metadata for gitcompatibledubpackage
Getting from http://localhost:4934/packages/gitcompatibledubpackage.json
Getting http://localhost:4934/packages/gitcompatibledubpackage.json...
Fetching gitcompatibledubpackage 1.0.2...
Acquiring package zip file
Downloading from 'http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip'
Storing http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip...
Download http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip 502 Bad Gateway (1.1)
Failed to download package gitcompatibledubpackage from http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip (Attempt 1 of 3)
Storing http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip...
Download http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip 502 Bad Gateway (1.1)
Failed to download package gitcompatibledubpackage from http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip (Attempt 2 of 3)
Storing http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip...
Download http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip 502 Bad Gateway (1.1)
Failed to download package gitcompatibledubpackage from http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip (Attempt 3 of 3)
Failed to download package gitcompatibledubpackage from http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip
Full exception: object.Exception@source/dub/packagesupplier.d(208): Failed to download package gitcompatibledubpackage from http://localhost:4934/packages/gitcompatibledubpackage/1.0.2.zip
----------------
source/dub/packagesupplier.d:208 void dub.packagesupplier.RegistryPackageSupplier.fetchPackage(dub.internal.vibecompat.inet.path.NativePath, immutable(char)[], dub.dependency.Dependency, bool) [0x762422]
source/dub/dub.d:766 dub.package_.Package dub.dub.Dub.fetch(immutable(char)[], const(dub.dependency.Dependency), dub.project.PlacementLocation, dub.dub.FetchOptions, immutable(char)[]) [0x74bafa]
source/dub/commandline.d:1235 int dub.commandline.FetchCommand.execute(dub.dub.Dub, immutable(char)[][], immutable(char)[][]) [0x70c427]
source/dub/commandline.d:252 int dub.commandline.runDubCommandLine(immutable(char)[][]) [0x704e8a]
source/app.d:14 _Dmain [0x6fc025]
========== -Output was ==========
[ERROR] :DUB should have retried download on server error multiple times, but only tried 1 times. command failed
```